enhancements/cluster-logging/cluster-logging-log-forwarding.md

```diff
@@ -0,0 +1,312 @@
+---
+title: cluster-logging-log-forwarding
+authors:
+  - "@jcantrill"
+reviewers:
+  - "@bparees"
+  - "@ewolinetz"
+  - "@jeremyeder"
+approvers:
+  - TBD
+creation-date: 2019-09-17
+last-updated: 2019-09-17
+status: implementable
+see-also:[]
+replaces:[]
+superseded-by:[]
+---
+
+# cluster-logging-log-forwarding
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Summary
+
+The purpose of log forwarding is to provide a declarative way by which adopters of cluster logging can ship
+container and node logs to destinations that are not necessarily managed by the OKD cluster logging infrastructure.
+Destinations are either on or off cluster endpoints such as the cluster logging provided Elasticsearch, an organization's Kafka message bus, or other log aggregation solutions (e.g. Splunk, Graylog).
+
+## Motivation
+
+Organizations desire to reuse their existing enterprise log solutions to store container logs.  Providing a declarative mechanism by which administrators define a log destination simplifies their operational burden.  The cluster logging stack is able to deploy a collector to each OKD node that is configured with the necessary permissions to collect the logs, add container metadata (e.g. labels) and ship them to the specified endpoint.
+
+
+
+### Goals
+The specific goals of this proposal are:
+
+* Provide an API that identifies log sources and their desired destination
+* Send logs to an Elasticsearch cluster not managed by the cluster logging infrastructure
+* Send logs to a fluentd instance not managed by the cluster logging infrastructure via fluentd's secureforward protocol
+* Allow deployment of log forwarding without deploying the entirety of the cluster logging infrastructure (e.g. Kibana, Elasticsearch)
+* Support TLS between the collector and destination if so configured
+* Re-use log forwarding to write logs to a cluster logging managed log aggregator (e.g. Elasticsearch)
+
```

Comment:
I expected to see “send logs to cloudwatch” as a target forward or easily accomplishable goal for a user.

Cloud users use cloudwatch.

Reply:
Cloudwatch, S3, etc. other log destinations are out of scope for this initial implementation.  We should add them to the future or planning sections.

Reply:
@smarterclayton looks like you already have the power simply by modifying the log driver: `--log-driver awslogs` [1]

[1] https://medium.com/attest-engineering/kubernetes-logs-to-aws-cloudwatch-with-fluentd-ede8d88a1b4e

Reply:
this would involve them using this feature to forward their logs from our fluentd to their fluentd, at which point they could configure their fluentd to forward anywhere they wanted, right?  or are you referring to a mechanism that would allow them to forward directly to aws from our fluentd?

Reply:
> this would involve them using this feature to forward their logs from our fluentd to their fluentd, at which point they could configure their fluentd to forward anywhere they wanted, right? or are you referring to a mechanism that would allow them to forward directly to aws from our fluentd?

i understand this configuration to modify the kublet logdriver to write directly to AWS cloudwatch completely bypassing use of fluentd.  Fluent currently tails logs from disk on the node.  Does this driver not suggest a direct write?  The post seems to conflate a couple ways of getting logs to cloudwatch

Reply:
I expect the design to cover how to forward to cloud watch, regardless of whether we implement it in 4.3

Reply:
There is a cloudwatch plugin for fluentd: https://github.com/yunomu/fluent-plugin-cloudwatch

Reply:
> I expect the design to cover how to forward to cloud watch, regardless of whether we implement it in 4.3

@smarterclayton if one can indeed write to cloud watch by modifying the kublet logdriver, would that not be preferred over introducing an intermediate service to do the job? I ask the question because in the cluster-logging case, its a matter of packaging the plugin and writing the code in the CLO where the the alternate case this falls on the responsibility of a different team who is configuring the node and outside the scope of this proposal

Reply:
Sounds like there are 3 options for covering this:

1) tell admins to modify the kubelet logdriver (i dunno how viable this is in OCP given how we control the nodes?)
2) tell admins to use secureforwrd to forward to their own fluentd instance which has the cloudwatch plugin configured
3) some future point at which we support direct forwarding to cloudwatch from our collector (whether that collector is fluentd, promtail, or something else).

@jcantrill for now i'd suggest you just denote (2), possibly under "non-goals", as in "it is not a goal to support specific log forwarding storage targets, as admins can simply forward to their own fluentd and then configure that fluentd to forward to any number of specific logstore targets"

Reply:
updated verbiage to address